src/main/scala/io/iohk/ethereum/db/dataSource/RocksDbDataSource.scala

```diff
@@ -0,0 +1,191 @@
+package io.iohk.ethereum.db.dataSource
+
+import java.util.concurrent.locks.ReentrantReadWriteLock
+
+import org.rocksdb._
+
+class RocksDbDataSource(private var db: RocksDB, private val rocksDbConfig: RocksDbConfig) extends DataSource {
+
+  /**
+    * This function obtains the associated value to a key, if there exists one.
+    *
+    * @param namespace which will be searched for the key.
+    * @param key       the key retrieve the value.
+    * @return the value associated with the passed key.
+    */
+  override def get(namespace: Namespace, key: Key): Option[Value] = {
+    RocksDbDataSource.dbLock.readLock().lock()
+    try {
+      val readOptions = new ReadOptions().setVerifyChecksums(rocksDbConfig.verifyChecksums)
+      Option(db.get(readOptions, (namespace ++ key).toArray))
+    } finally {
+      RocksDbDataSource.dbLock.readLock().unlock()
+    }
+  }
+
+  /**
+    * This function obtains the associated value to a key, if there exists one. It assumes that
+    * caller already properly serialized key. Useful when caller knows some pattern in data to
+    * avoid generic serialization.
+    *
+    * @param key the key retrieve the value.
+    * @return the value associated with the passed key.
+    */
+  override def getOptimized(key: Array[Byte]): Option[Array[Byte]] = {
+    RocksDbDataSource.dbLock.readLock().lock()
+    try {
+      val readOptions = new ReadOptions().setVerifyChecksums(rocksDbConfig.verifyChecksums)
+      Option(db.get(readOptions, key))
+    } finally {
+      RocksDbDataSource.dbLock.readLock().unlock()
+    }
+
+  }
+
+  /**
+    * This function updates the DataSource by deleting, updating and inserting new (key-value) pairs.
+    *
+    * @param namespace from which the (key-value) pairs will be removed and inserted.
+    * @param toRemove  which includes all the keys to be removed from the DataSource.
+    * @param toUpsert  which includes all the (key-value) pairs to be inserted into the DataSource.
+    *                  If a key is already in the DataSource its value will be updated.
+    * @return the new DataSource after the removals and insertions were done.
+    */
+  override def update(namespace: Namespace, toRemove: Seq[Key], toUpsert: Seq[(Key, Value)]): DataSource = {
+    RocksDbDataSource.dbLock.readLock().lock()
+    try {
+      val batch = new WriteBatch()
+      val writeOptions = new WriteOptions().setSync(rocksDbConfig.synchronousWrites)
```

Comment:
All rockDb objects are backed by c++ , and you need to call `close` after using them (https://github.com/facebook/rocksdb/wiki/RocksJava-Basics#memory-management). So even `WriteOptions ` object need to be closed, thats way ethereumj  uses `try-with-resources` for all alocated objects (https://github.com/ethereum/ethereumj/blob/develop/ethereumj-core/src/main/java/org/ethereum/datasource/rocksdb/RocksDbDataSource.java#L265) as all rocks db object implement `AutoClosableInterface`. 
Unfortunatly scala does not support this idiom, although it is possible to implement it yourself (check out https://medium.com/@dkomanov/scala-try-with-resources-735baad0fd7d). 
Maybe it is worth implementing such util and using it here ?